dmoj/cptbox/ptdebug_x64.cpp

```diff
@@ -2,96 +2,129 @@
 #define _BSD_SOURCE
 #include "ptbox.h"
 
-#ifdef HAS_DEBUGGER_X64
+#ifdef __amd64__
+#include <elf.h>
 #include <errno.h>
 #include <stdio.h>
 #include <stdlib.h>
 #include <string.h>
 #include <sys/ptrace.h>
+#include <sys/uio.h>
 
-#define R15 0
-#define R14 1
-#define R13 2
-#define R12 3
-#define RBP 4
-#define RBX 5
-#define R11 6
-#define R10 7
-#define R9 8
-#define R8 9
-#define RAX 10
-#define RCX 11
-#define RDX 12
-#define RSI 13
-#define RDI 14
-#define ORIG_RAX 15
-#define RIP 16
-#define CS 17
-#define EFLAGS 18
-#define RSP 19
-#define SS 20
-#define FS_BASE 21
-#define GS_BASE 22
-#define DS 23
-#define ES 24
-#define FS 25
-#define GS 26
+#ifndef __X32_SYSCALL_BIT
+#   define __X32_SYSCALL_BIT 0x40000000UL
+#endif
 
-int pt_debugger_x64::syscall() {
-    return (int) peek_reg(ORIG_RAX);
+bool pt_debugger::supports_abi(int abi) {
+    switch (abi) {
+        case PTBOX_ABI_X86:
+        case PTBOX_ABI_X64:
+        case PTBOX_ABI_X32:
+            return true;
+    }
+    return false;
 }
 
-void pt_debugger_x64::syscall(int id) {
-#if PTBOX_FREEBSD
-    poke_reg(RAX, id);
-#else
-    poke_reg(ORIG_RAX, id);
-#endif
+void pt_debugger::pre_syscall() {
+    struct iovec iovec;
+    iovec.iov_base = &regs;
+    iovec.iov_len = sizeof regs;
+    if (ptrace(PTRACE_GETREGSET, tid, NT_PRSTATUS, &iovec))
+        perror("ptrace(PTRACE_GETREGSET)");
```

Comment:
I don't think `perror` is harsh enough for signalling that we're about to send up garbage.